ci-matrix.py

```diff
@@ -0,0 +1,191 @@
+# /// script
+# requires-python = ">=3.13"
+# dependencies = [
+#     "packaging",
+#     "pyyaml",
+# ]
+# ///
+
+import argparse
+import json
+from typing import Any, Dict, List, Optional, Set
+
+import yaml
+from packaging.version import Version
+
+CI_TARGETS_YAML = "ci-targets.yaml"
+
+
+def meets_conditional_version(version: str, min_version: str) -> bool:
+    return Version(version) >= Version(min_version)
+
+
+def parse_labels(labels: Optional[str]) -> Dict[str, Set[str]]:
+    """Parse labels into a dict of category filters."""
+    if not labels:
+        return {}
+
+    result: Dict[str, Set[str]] = {
+        "platform": set(),
+        "python": set(),
+        "build": set(),
+        "arch": set(),
+        "libc": set(),
+    }
+
+    for label in labels.split(","):
+        label = label.strip()
+        if not label or ":" not in label:
+            continue
+        category, value = label.split(":", 1)
+        if category in result:
+            result[category].add(value)
+
+    return result
+
+
+def should_include_entry(entry: Dict[str, str], filters: Dict[str, Set[str]]) -> bool:
+    """Check if an entry satisfies the label filters."""
+    if filters.get("platform") and entry["platform"] not in filters["platform"]:
+        return False
+
+    if filters.get("python") and entry["python"] not in filters["python"]:
+        return False
+
+    if filters.get("arch") and entry["arch"] not in filters["arch"]:
+        return False
+
+    if filters.get("libc") and entry.get("libc") not in filters["libc"]:
+        return False
+
+    if filters.get("build"):
+        build_options = set(entry.get("build_options", "").split("+"))
+        if not all(f in build_options for f in filters["build"]):
+            return False
+
+    return True
+
+
+def generate_matrix_entries(
+    config: Dict[str, Any],
+    platform_filter: Optional[str] = None,
+    label_filters: Optional[Dict[str, Set[str]]] = None,
+) -> List[Dict[str, str]]:
+    matrix_entries = []
+
+    for platform, platform_config in config.items():
+        if platform_filter and platform != platform_filter:
+            continue
+
+        for target_triple, target_config in platform_config.items():
+            add_matrix_entries_for_config(
+                matrix_entries,
+                target_triple,
+                target_config,
+                platform,
+            )
+
+    # Apply label filters if present
+    if label_filters:
+        matrix_entries = [
+            entry
+            for entry in matrix_entries
+            if should_include_entry(entry, label_filters)
+        ]
+
+    return matrix_entries
+
+
+def add_matrix_entries_for_config(
+    matrix_entries: List[Dict[str, str]],
```

Comment:
Can we just use builtins for this rather than `List`, `Dict`, etc.? 3.13 is already the minimum. (You need Ruff!)

Reply:
Ooh yeah happy to do that. I didn't actually set the minimum, uv added it without me noticing, but regardless.

Also I don't think that rule is selected for Ruff — there is a `ruff.toml` here!

Reply:
Fixed with `ruff check --select UP006 --target-version py313 ci-matrix.py --fix`